src/geos_types.jl

```diff
@@ -323,51 +348,78 @@ end
-    is3d === GeoInterface.is3d(pt2) || return false
-    GeoInterface.getcoord(pt1,1) ≅ GeoInterface.getcoord(pt2,1) || return false
-    GeoInterface.getcoord(pt1,2) ≅ GeoInterface.getcoord(pt2,2) || return false
-    if is3d
-        GeoInterface.getcoord(pt1,3) ≅ GeoInterface.getcoord(pt2,3) || return false
+    if has_coord_seq(geo1)
+        return compare_coord_seqs(cmp, geo1, geo2, ctx)
+    else
+        ng1 = ngeom(geo1)
+        ng2 = ngeom(geo2)
+        ng1 == ng2 || return false
+        for i in 1:ng1
+            compare(cmp, getgeom(geo1, i), getgeom(geo2, i), ctx) || return false
+        end
     end
     return true
 end
 
-function _norm(x,y,z)
-    return sqrt(x^2 + y^2 + z^2)
+npoints(pt::Point) = 1
+npoints(geo::HasCoordSeq) = GeoInterface.ngeom(geo)
+
+function compare_coord_seqs(cmp, geo1, geo2, ctx)
+    ncoords1 = GeoInterface.ncoord(geo1)
+    ncoords2 = GeoInterface.ncoord(geo2)
+    ncoords1 == ncoords2 || return false
+    ncoords1 == 0 && return true
+    np1 = npoints(geo1)
+    np2 = npoints(geo2)
+    np1 == np2 || return false
+    coords1 = Vector{Float64}(undef, ncoords1)
+    coords2 = Vector{Float64}(undef, ncoords1)
+    for i in 1:np1
+        coordinates!(coords1, geo1, i, ctx)
+        coordinates!(coords2, geo2, i, ctx)
+        cmp(coords1, coords2) || return false
+    end
+    return true
 end
 
-function compare(cmp::IsApprox, pt1::Point, pt2::Point)::Bool
-    is3d = GeoInterface.is3d(pt1)
-    is3d === GeoInterface.is3d(pt2) || return false
-    x1 = GeoInterface.getcoord(pt1,1)
-    x2 = GeoInterface.getcoord(pt2,1)
-    y1 = GeoInterface.getcoord(pt1,2)
-    y2 = GeoInterface.getcoord(pt2,2)
-    if is3d
-        z1 = GeoInterface.getcoord(pt1,3) 
-        z2 = GeoInterface.getcoord(pt2,3)
-    else
-        z1 = 0.0
-        z2 = 0.0
+function compare_coord_seqs(cmp::IsApprox, geo1, geo2, ctx)
+    ncoords1 = GeoInterface.ncoord(geo1)
+    ncoords2 = GeoInterface.ncoord(geo2)
+    ncoords1 == ncoords2 || return false
+    ncoords1 == 0 && return true
+    @assert ncoords1 in 2:3
+    ncoords1 == ncoords2 || return false
+    np1 = npoints(geo1)
+    np2 = npoints(geo2)
+    np1 == np2 || return false
+    coords1 = Vector{Float64}(undef, ncoords1)
+    coords2 = Vector{Float64}(undef, ncoords1)
+    # isapprox of two vectors is calculated using their euclidean norms and the norm of their difference
+    # we compute (the squares) of these norms in an allocation free way
+    s1 = 0.0
+    s2 = 0.0
+    s12 = 0.0
+    for i in 1:np1
+        coordinates!(coords1, geo1, i, ctx)
+        coordinates!(coords2, geo2, i, ctx)
+        if ncoords1 == 2
+            x1,y1 = coords1
+            x2,y2 = coords2
+            s12 += (x1 - x2)^2 + (y1 - y2)^2
+        else
+            x1,y1,z1 = coords1
+            x2,y2,z2 = coords2
+            s12 += (x1 - x2)^2 + (y1 - y2)^2 + (z1 - z2)^2
+        end
+        s1 += sum(abs2, coords1)
+        s2 += sum(abs2, coords2)
```

Comment:
Just for my understanding: where does `abs2` here come from? I missed the test case which covers the implementation of `compare_coord_seqs(cmp::IsApprox, geo1, geo2, ctx)` to step through it

Reply:
* Given a geometries defined by points `pts1 = [pt1, pt2, pt3...]` and `pts2 = [pt1', pt2', ...]` we calculate 
`s1 = norm(pts1)^2` and `s2 = norm(pts2)^2` and `s12 = norm(pts1 - pts2)^2`. These are sums of squares, that can be computed using `s1 += sum(abs2, pt)` or `s1 += x1^2 + y1^2`.

* There are lots of `isapprox(geo1, geo2)` tests which hit this implementation at the end of the day. Is this what you mean?

Reply:
Note that these tests existed before the PR, this PR only makes existing functionality faster. So you don't see all tests in the diff.

Reply:
Yeah [codecov](https://app.codecov.io/gh/JuliaGeo/LibGEOS.jl/pull/159/blob/src/geos_types.jl) does indicate coverage. My bad, I was reading `abs2` as a variable when it's a function:

```julia
help?> abs2
search: abs2 abs abspath abstract AbstractSet 'abstract type' AbstractChar

  abs2(x)

  Squared absolute value of x.

  Examples
  ≡≡≡≡≡≡≡≡≡≡

  julia> abs2(-3)
  9

julia> sum(abs2, [1,2])
5
```

The implementation makes sense to me now.